### PR TITLE
*: add label check for create placement policy

### LINKED ISSUE #26580: Support named PLACEMENT POLICY
This is a sub-task of Placement Rules in SQL: https://github.com/pingcap/tidb/issues/18030

Support the ability to `CREATE PLACEMENT POLICY` with the same options available in `CREATE TABLE|ALTER TABLE` context.

The following subtasks are known:

* [x] Parser: {CREATE|ALTER} PLACEMENT POLICY placement_options https://github.com/pingcap/parser/pull/1313
* [x] Parser: SHOW CREATE PLACEMENT POLICY x, DROP PLACEMENT POLICY [IF EXISTS] x; https://github.com/pingcap/parser/pull/1295 https://github.com/pingcap/parser/pull/1298
* [x] Design table in mysql.* for persisting placement policies. https://github.com/pingcap/tidb/pull/27251
* [x] Ensure {CREATE|ALTER|DROP} PLACEMENT POLICY requires the PLACEMENT_ADMIN privilege https://github.com/pingcap/tidb/issues/27603
* [x] #27714
* [ ] ALTER PLACEMENT POLICY should update placement rules for all tables in scope in PD.
* [ ] DROP PLACEMENT POLICY should clear the placement rules for all tables in scope in PD, and reset the `SHOW CREATE TABLE` info to no longer have the placement policy named.

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: partial close #26580

### What is changed and how it works?

What's Changed:
when create placement policy name ???_constraints=[`label1`=vakue], constraints=[`label2`=value], the common constraints is actually appended to the previous constraints. 

This way, after the check of compatibility, we should also check the `label existence` in the storage via access PD api, which is already included in `show placement labels`

### Check List 

Tests 

- [x] Manual test (add detailed scripts or steps below)

```
script
1: start a real cluster:
tiup playground 5.1.1 --tiflash 0  --db 2 --db.binpath  ~/go/src/github.com/pingcap/tidb/bin/tidb-server

2: restart the tikv node with labels:
/tikv-server --addr=127.0.0.1:20160 --advertise-addr=127.0.0.1:20160 --status-addr=127.0.0.1:20180 --pd=http://127.0.0.1:2379 --config=/home/arenatlx/.tiup/data/Shjeocq/tikv-0/tikv.toml --data-dir=/home/arenatlx/.tiup/data/Shjeocq/tikv-0/data --log-file=/home/arenatlx/.tiup/data/Shjeocq/tikv-0/tikv.log --labels zone=beijing,rack=huawei,host=local,disk=ssd

3: test show placement label can work:
mysql> show placement labels;
+------+-------------+
| Key  | Values      |
+------+-------------+
| disk | ["ssd"]     |
| host | ["local"]   |
| rack | ["huawei"]  |
| zone | ["beijing"] |
+------+-------------+
4 rows in set (0.01 sec)

4: test create placement policy with vaild labels:
mysql> create placement policy x followers 2 follower_constraints="[+disk=ssd]" constraints="[+zone=beijing]";
Query OK, 0 rows affected (0.53 sec)

5: test create placement policy with invalid labels:
mysql> create placement policy x1 followers 2 follower_constraints="[+disk=ssd]" constraints="[+zone1=beijing]";
ERROR 1105 (HY000): Placement label zone1 doesn't exist in the storage
```

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by xhebox — COMMENTED** [review]
LGTM, but I think it is not strictly necessary. And it adds another tricky SQL usage.

PD will handle unexisted labels, in fact. In another word, it will report error, if what we posted can not be satisfied. Futhermore, what if a user add the unexisted label later?

My opionion is that we don't need to go too far about the validation of placement policy when creating, but binding. It won't affect correctness even if we don't validate the policy.

A strict validation will be done when binding/alter anyway. One just can not use incorrect policies.

**Comment by lcwangchao** [comment]
I think you should update this branch because `alter placement policy is merged`

File: ddl/placement_policy.go [hunk 1]
```diff
@@ -65,15 +67,62 @@ func onCreatePlacementPolicy(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64
...
+	}
+	err = b.CheckLabels(distinctLabels)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (w *worker) checkPolicyConstraints(info *placementpolicy.PolicyInfo) error {
+	var totalRules []*placement.Rule
```
Comment (lcwangchao) [inline ddl/placement_policy.go:112]: Many logics are duplicated with pr: #27814 . Can we reuse the code after one committed?

**Comment by AilinKid** [comment]
> LGTM, but I think it is not strictly necessary. And it adds another tricky SQL usage.
> 
> PD will handle unexisted labels, in fact. In another word, it will report error, if what we posted can not be satisfied. Futhermore, what if a user add the unexisted label later?
> 
> My opionion is that we don't need to go too far about the validation of placement policy when creating, but binding. It won't affect correctness even if we don't validate the policy.
> 
> A strict validation will be done when binding/alter anyway. One just can not use incorrect policies.

make sense, should we need these check when `create table with policy/direct options`, or just let the PD to validate it?

**Comment by xhebox** [comment]
> 
> 
> > LGTM, but I think it is not strictly necessary. And it adds another tricky SQL usage.
> > PD will handle unexisted labels, in fact. In another word, it will report error, if what we posted can not be satisfied. Futhermore, what if a user add the unexisted label later?
> > My opionion is that we don't need to go too far about the validation of placement policy when creating, but binding. It won't affect correctness even if we don't validate the policy.
> > A strict validation will be done when binding/alter anyway. One just can not use incorrect policies.
> 
> make sense, should we need these check when `create table with policy/direct options`?, or just let the PD to validate it?

I think using `checkPolicyValidation` or `NewBundleFromOptions` to eliminate some simple/basic errors will improve the user experience, so that one knows the error immediately before binding. Any other complex validation does not seem reasonable.

**Comment by AilinKid** [comment]
> > > LGTM, but I think it is not strictly necessary. And it adds another tricky SQL usage.
> > > PD will handle unexisted labels, in fact. In another word, it will report error, if what we posted can not be satisfied. Futhermore, what if a user add the unexisted label later?
> > > My opionion is that we don't need to go too far about the validation of placement policy when creating, but binding. It won't affect correctness even if we don't validate the policy.
> > > A strict validation will be done when binding/alter anyway. One just can not use incorrect policies.
> > 
> > 
> > make sense, should we need these check when `create table with policy/direct options`?, or just let the PD to validate it?
> 
> I think using `checkPolicyValidation` or `NewBundleFromOptions` to eliminate some simple/basic errors will improve the user experience, so that one knows the error immediately before binding. Any other complex validation does not seem reasonable.

Right, I'll close this pr, thx @lcwangchao @xhebox

**Pull request closed without merging**